File: leetcode-patterns/diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

let diameter = 0;

/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  diameter = 0;
  heightOfBinaryTree(root);
  return diameter;
};

var heightOfBinaryTree = function (root) {
  if (root === null) {
    return -1;
  }

  const leftHeight = heightOfBinaryTree(root.left);
  const rightHeight = heightOfBinaryTree(root.right);

  diameter = Math.max(diameter, leftHeight + rightHeight + 2);

  return Math.max(leftHeight, rightHeight) + 1;
};

File: leetcode/leetcode-patterns/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def __init__(self):
    self.d = 0
  
  def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
    self.height(root)
    return self.d
  
  def height(self, node: TreeNode) -> int:
    if node is None:
      return -1
    lh = self.height(node.left)
    rh = self.height(node.right)
    self.d = max(self.d, lh + rh + 2)
    return max(lh, rh) + 1
    
    
        